### ISSUE TITLE
[RFC] `torchvision.transforms` revamp

### ISSUE THREAD

**Issue by pmeier** [pytorch/vision#5754]
### 🚀 The feature

*Note: To track the progress of the project check out [this board](https://github.com/pytorch/vision/projects/5).*

The current transforms in the `torchvision.transforms` name space are limited to images. This makes it hard to use them for tasks that require the transform to be applied not only to the input image, but also to the target. For example, in object detection, resizing or cropping the input image also affects the bounding boxes. 

This projects aims to resolve this by providing transforms that can handle the full sample possibly including images, bounding boxes, segmentation masks, and so on without the need for user interference. The implementation of this project happens in the `torchvision.prototype.transforms` namespace. For example:

```py
from torchvision.prototype import features, transforms

# this will be supplied by a dataset from torchvision.prototype.datasets
image = features.EncodedImage.from_path("test/assets/fakedata/logos/rgb_pytorch.png")
label = features.Label(0)

transform = transforms.Compose(
    transforms.DecodeImage(),
    transforms.RandomHorizontalFlip(p=1.0),
    transforms.Resize((100, 300)),
)

transformed_image1 = transform(image)
transformed_image2, transformed_label = transform(image, label)

# whether we call the transform with label or not has no effect on the image transform
assert transformed_image1.eq(transformed_image2).all()
# the transform is a no-op for labels
assert transformed_label.eq(label).all()
```

![before_image](https://user-images.githubusercontent.com/6849766/161979289-b580ef3c-b947-4f9d-be3e-6b05659736aa.png) ![after_image](https://user-images.githubusercontent.com/6849766/161979304-2f3a30d0-991f-4ad1-b091-61600d49fdb3.png)

```py
# this will be supplied by a dataset from torchvision.prototype.datasets
bounding_box = features.BoundingBox(
    [60, 30, 15, 15], format=features.BoundingBoxFormat.CXCYWH, image_size=(100, 100)
)

transformed_image, transformed_bounding_box = transform(image, bounding_box)
```

![before_image_and_bounding_box](https://user-images.githubusercontent.com/6849766/161932345-652f3012-7427-44b4-aed8-952249db3345.png) ![after_image_and_bounding_box](https://user-images.githubusercontent.com/6849766/161932375-47ca2a19-ff36-40c4-af78-579351466184.png)

```py
# this will be supplied by a dataset from torchvision.prototype.datasets
segmentation_mask = torch.zeros((100, 100), dtype=torch.bool)
segmentation_mask[24:36, 55:66] = True
segmentation_mask = features.SegmentationMask(segmentation_mask)

transformed_image, transformed_segmentation_mask = transform(image, segmentation_mask)
```

![before_image_and_segmentation_mask](https://user-images.githubusercontent.com/6849766/161980409-58afc8ce-9b4c-4c47-937d-f3ed551218e5.png) ![after_image_and_segmentation_mask](https://user-images.githubusercontent.com/6849766/161980425-e6e8f4d5-18f9-4f83-a3e4-8a6a19de6805.png)

### Classification

- [ ] Port transforms from `torchvision.transforms` #5520
- [x] Port transforms from the classification references #5780

### Detection

- [ ] New kernels for bounding boxes #5514
- [ ] Port transforms from the detection references #5542

### Segmentation

- [ ] New kernels for segmentation masks #5782
- [ ] Port transforms from segmentation references. This will automatically be achieved with the completion above, since there are no special [segmentation transforms](https://github.com/pytorch/vision/blob/main/references/segmentation/transforms.py) in our references. If the respective kernels are implemented, the transforms will handle the segmentation case out of the box.

### Other

- [ ] #5626

cc @vfdev-5 @datumbox

**Comment by vfdev-5** [comment]
I'm closing this tracker as done
